backend/sirius-web-spring-collaborative-diagrams/src/main/java/org/eclipse/sirius/web/spring/collaborative/diagrams/DiagramService.java

```diff
@@ -109,4 +137,45 @@ public Diagram create(DiagramCreationParameters parameters) {
                 .map(Diagram.class::cast);
         // @formatter:on
     }
+
+    private boolean containsUnsyncDescriptions(DiagramDescription diagramDescription) {
+        return diagramDescription.getNodeDescriptions().stream().anyMatch(this::hasOneNodeDescriptionUnsynchronized);
+    }
+
+    private boolean hasOneNodeDescriptionUnsynchronized(NodeDescription nodeDescription) {
+        boolean isUnsynchronised = !nodeDescription.isSynchronised();
+        if (!isUnsynchronised) {
+            isUnsynchronised = nodeDescription.getBorderNodeDescriptions().stream().anyMatch(this::hasOneNodeDescriptionUnsynchronized);
+            if (!isUnsynchronised) {
+                isUnsynchronised = nodeDescription.getChildNodeDescriptions().stream().anyMatch(this::hasOneNodeDescriptionUnsynchronized);
+            }
+        }
+        return isUnsynchronised;
+    }
+
+    private Map<String, Object> getSemanticElements(Diagram diagram, IEditingContext editingContext) {
+        Map<String, Object> result = new HashMap<>();
+        for (String targetId : this.getTargetIds(diagram)) {
+            Optional<Object> optionalObject = this.objectService.getObject(editingContext, targetId);
+            if (optionalObject.isPresent()) {
+                result.put(targetId, optionalObject.get());
+            }
+        }
+        return result;
+    }
+
+    private List<String> getTargetIds(Diagram diagram) {
+        List<String> result = new ArrayList<>();
+        result.add(diagram.getTargetObjectId());
+        diagram.getNodes().forEach(node -> result.addAll(this.getTargetIds(node)));
+        return result;
+    }
+
+    private List<String> getTargetIds(Node node) {
+        List<String> result = new ArrayList<>();
+        result.add(node.getTargetObjectId());
+        node.getBorderNodes().stream().map(this::getTargetIds).flatMap(List::stream).forEach(result::add);
+        node.getChildNodes().stream().map(this::getTargetIds).flatMap(List::stream).forEach(result::add);
+        return result;
+    }
```

Comment:
This seems to allocate quite a lot of intermediate lists.

 How about something like (untested):

```java
    private List<String> getTargetIds(Diagram diagram) {
        List<String> result = new ArrayList<>();
        result.add(diagram.getTargetObjectId());
        diagram.getNodes().forEach(node -> this.collectTargetIds(node, result));
        return result;
    }

    private void collectTargetIds(Node node, List<String> result) {
        result.add(node.getTargetObjectId());
        node.getBorderNodes().stream().map(borderNode -> this.collectTargetIds(borderNode, result));
        node.getChildNodes().stream().map(childNode -> this.collectTargetIds(childNode, result));
    }
```

Side-note:  maybe `Diagram` should have something like `forEachNode(Consumer<Node>)`.